### ISSUE TITLE
SDK support for Proxy requiring Kerberos auth

### ISSUE THREAD

**Issue by PankajSAgarwal** [aws/aws-sdk-java-v2#4767]
### Describe the feature

AWS SDK2 Java does not provide ability to authenticate to AWS STS via proxy that requires Kerberos Authentication.

### Use Case

Need to connect from on-prem to AWS STS API via proxy which requires Kerberos authentication .

We have run into a limitation of the AWS SDK ApacheHttpClient which doesn’t allow for customization of the builder:
[https://github.com/aws/aws-sdk-java-v2/blob/2.20.156/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java#L153 ](https://github.com/aws/aws-sdk-java-v2/blob/2.20.156/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java#L153)
which is required for us to configure the setDefaultAuthSchemeRegistry and setDefaultCredentialsProvider:

### Proposed Solution

As a workaround we implemented a forked implementation of the AWS `ApacheHttpClient` to override the AuthSchemeProvider which will consider `canonicalHostName` of proxy by default for kerberos authemtication .

We added below code snippet to the builder at the following line https://github.com/aws/aws-sdk-java-v2/blob/2.20.156/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java#L153
```java #2
Registry<AuthSchemeProvider> authSchemeProviderRegistry = configuration.authSchemeProviderRegistry;
        if (authSchemeProviderRegistry == null) {
            authSchemeProviderRegistry = RegistryBuilder.<AuthSchemeProvider>create()
                    .register(AuthSchemes.BASIC, new BasicSchemeFactory())
                    .register(AuthSchemes.DIGEST, new DigestSchemeFactory())
                    .register(AuthSchemes.NTLM, new NTLMSchemeFactory())
                    .register(AuthSchemes.SPNEGO, new SPNegoSchemeFactory(true, false))
                    .register(AuthSchemes.KERBEROS, new KerberosSchemeFactory(true, false)).build();
        }
        builder.setDefaultAuthSchemeRegistry(authSchemeProviderRegistry);
``` 
We also had to use this system property in our application to make the kerberos negotiator work with AWS SDK Java 2 client

```bash
System.setProperty("javax.security.auth.useSubjectCredsOnly", "false");
```

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### AWS Java SDK version used

aws sdk java v2

### JDK version used

17 and 21

### Operating System and version

Windows Server 2016 Standard/ Linux

**Comment by debora-ito** [comment]
@PankajSAgarwal I believe you submitted a Premium Support case with the same ask. As we said in the case, we've added this to our backlog.

For a similar feature request, we exposed the Apache HttpRoutePlanner attribute in the SDK ApacheHttpClient builder ([javadocs](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/http/apache/ApacheHttpClient.Builder.html#httpRoutePlanner(org.apache.http.conn.routing.HttpRoutePlanner))). It's not exactly the feature you are asking for, but maybe you can use HttpRoutePlanner to redirect the flow to the authentication server?

**Comment by PankajSAgarwal** [comment]
@debora-ito  , yes that is correct , I was advised by support perosanl to raise a feature request for the same

**Comment by debora-ito** [comment]
> I was advised by support personal on the case to raise a feature request for the same on Github.

Understood. 

What about Apache HttpRoutePlanner, would it work for you?

**Comment by PankajSAgarwal** [comment]
> > I was advised by support personal on the case to raise a feature request for the same on Github.
> 
> Understood.
> 
> What about Apache HttpRoutePlanner, would it work for you?

HttpRoutePlanner will not work as well , HttpRoutePlanner can direct me to proxy and i can use basic authentication , but my requirement is to be able add kerberos Auth scheme for authentication to proxy , which HttpRoutePlanner doesn't seem to be supporting .

**Comment by Fred1155** [comment]
Hi, I’m trying to investigating on this issue. Could you please provide us your forked implementation (without any your sensitive information)? It would be very helpful if we know what fields you config into the ApacheHttpClient.

**Comment by debora-ito** [comment]
Support for this was added via https://github.com/aws/aws-sdk-java-v2/pull/5874 and released in Java SDK version [2.30.20](https://github.com/aws/aws-sdk-java-v2/blob/master/CHANGELOG.md#23020-2025-02-13).

Now it's possible to configure the AuthSchemeProviderRegistry in the ApacheHttpClient builder, and KERBEROS is one of the AuthScheme options. Check our [unit tests](https://github.com/aws/aws-sdk-java-v2/blob/9f735a10ba218860d15eeb4a8fcb6105ca04e92e/http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/ApacheHttpClientAuthRegistryTest.java#L123-L156) for some quick examples of use.

**Comment by warlockza** [comment]
Was there any plans to support the Netty library for the same?

**Comment by dagnir** [comment]
@warlockza Created #7033 to track for Netty.
